app/addons/dataimporter/stores.js

```diff
@@ -0,0 +1,382 @@
+
+      // the following object shows all of the options available for papaparse
+      // some are defaulted to undefined
+      // this is from http://papaparse.com/docs#config
+      return {
+        delimiter : '',  // auto-detect
+        newline: '',  // auto-detect
+        header: true,
+        dynamicTyping: true,
+        preview: 0,
+        encoding: '',
+        worker: true, //true = page doesn't lock up
+        comments: false,
+        complete: function (results) {
+          this.loadingComplete(results);
+          this._changingConfig = false;
+        }.bind(this),
+        error: function () {
+          var msg1 = 'There was an error while parsing the file.',
+              msg2 = 'Please try again.';
+
+          this.goToErrorScreen('', [msg1, msg2]);
+        }.bind(this),
+        download: false,
+        skipEmptyLines: false,
+        chunk: undefined, //define function for streaming
+        beforeFirstChunk: undefined,
+      };
+    },
+
+    loadDataIntoDatabase: function (createNewDB, targetDB) {
+      this._loadingInDBInProgress = true;
+      this.triggerChange();
+      if (createNewDB) {
+        if (this.databaseIsNew(targetDB)) {
+          var msg1 = 'The database ' + targetDB + ' already exists.';
+          this.goToErrorScreen('', [msg1]);
+        }
+        this.loadIntoNewDB(targetDB);
+      } else {
+        this.loadDataIntoTarget(targetDB);
+      }
+    },
+
+    databaseIsNew: function (targetDB) {
+      return _.some(this._all_dbs, targetDB);
+    },
+
+    showErrorScreen: function () {
+      return this._showErrorScreen;
+    },
+
+    getErrorMsg: function () {
+      return this._errorMessageArray;
+    },
+
+    goToErrorScreen: function (resp, messageArray) {
+      this._loadingInDBInProgress = false;
+      this._showErrorScreen = true;
+      if (resp) {
+        messageArray.push(resp);
+      }
+      this._errorMessageArray.unshift(messageArray);
+      this.triggerChange();
+    },
+
+    loadIntoNewDB: function (targetDB) {
+      $.ajax({
+        url: FauxtonAPI.urls('databaseBaseURL', 'server', targetDB),
+        xhrFields: { withCredentials: true },
+        contentType: 'application/json; charset=UTF-8',
+        method: 'PUT'
+      }).then(function (resp) {
+        this.loadDataIntoTarget(targetDB);
+      }.bind(this), function (resp) {
+        this.importFailed();
+      }.bind(this));
+    },
+
+    loadDataIntoTarget: function (targetDB) {
+      var loadURL = FauxtonAPI.urls('document', 'server', targetDB, '_bulk_docs');
+      _.each(this._chunkedData, function (data, i) {
+        var payload = JSON.stringify({ 'docs': data });
+        var prettyprint = JSON.stringify({ 'docs': data }, null, 2);
+        $.ajax({
+          url: loadURL,
+          xhrFields: { withCredentials: true },
+          contentType: 'application/json; charset=UTF-8',
+          method: 'POST',
+          data: payload
+        }).then(function (resp) {
+          i++;
```

Comment:
every ajax request returns a promise. you can push all ajax requests onto an array, so you have an array of promises and then use `FauxtonAPI.when` which will resolve when all promises in the array resolve - it is slightly more elegant than manually counting and has less side effects.

example: https://github.com/apache/couchdb-fauxton/blob/5de1fe6682fc5a9dee44022498357febd73c3600/app/addons/cors/actions.js#L22